spanner/tests/unit/test_keyset.py

```diff
@@ -30,49 +33,68 @@ def test_ctor_no_start_no_end(self):
+    def test_ctor_end_open_and_end_closed(self):
         KEY_1 = [u'key_1']
         KEY_2 = [u'key_2']
         with self.assertRaises(ValueError):
             self._make_one(end_open=KEY_1, end_closed=KEY_2)
 
-    def test_ctor_w_only_start_open(self):
+    def test_ctor_conflicting_start(self):
         KEY_1 = [u'key_1']
-        krange = self._make_one(start_open=KEY_1)
-        self.assertEqual(krange.start_open, KEY_1)
-        self.assertEqual(krange.start_closed, None)
-        self.assertEqual(krange.end_open, None)
-        self.assertEqual(krange.end_closed, None)
+        with self.assertRaises(ValueError):
+            self._make_one(start_open=[], start_closed=[], end_closed=KEY_1)
 
-    def test_ctor_w_only_start_closed(self):
+    def test_ctor_conflicting_end(self):
+        KEY_1 = [u'key_1']
+        with self.assertRaises(ValueError):
+            self._make_one(start_open=KEY_1, end_open=[], end_closed=[])
+
+    def _check_unused_keys(self, key_range, used_keys):
+        key_types = ('start_closed',
+                     'start_open',
+                     'end_closed',
+                     'end_open')
+        for key_type in key_types:
+            if key_type not in used_keys:
+                self.assertFalse(key_range.to_pb().HasField(key_type))
+
+    def test_ctor_single_key_start_closed(self):
         KEY_1 = [u'key_1']
         krange = self._make_one(start_closed=KEY_1)
-        self.assertEqual(krange.start_open, None)
         self.assertEqual(krange.start_closed, KEY_1)
-        self.assertEqual(krange.end_open, None)
-        self.assertEqual(krange.end_closed, None)
+        self.assertEqual(krange.end_closed, [])
+        self.assertIsNone(krange.start_open)
+        self.assertIsNone(krange.end_open)
 
-    def test_ctor_w_only_end_open(self):
+    def test_ctor_single_key_start_open(self):
         KEY_1 = [u'key_1']
-        krange = self._make_one(end_open=KEY_1)
-        self.assertEqual(krange.start_open, None)
-        self.assertEqual(krange.start_closed, None)
-        self.assertEqual(krange.end_open, KEY_1)
-        self.assertEqual(krange.end_closed, None)
+        krange = self._make_one(start_open=KEY_1)
+        self.assertEqual(krange.start_open, KEY_1)
+        self.assertEqual(krange.end_closed, [])
+        self.assertIsNone(krange.start_closed)
+        self.assertIsNone(krange.end_open)
 
-    def test_ctor_w_only_end_closed(self):
+    def test_ctor_single_key_end_closed(self):
         KEY_1 = [u'key_1']
         krange = self._make_one(end_closed=KEY_1)
-        self.assertEqual(krange.start_open, None)
-        self.assertEqual(krange.start_closed, None)
-        self.assertEqual(krange.end_open, None)
         self.assertEqual(krange.end_closed, KEY_1)
+        self.assertEqual(krange.start_closed, [])
+        self.assertIsNone(krange.start_open)
+        self.assertIsNone(krange.end_open)
+
+    def test_ctor_single_key_end_open(self):
```

Comment:
1. I don't think it's necessary to test **all four** possible "single input" cases
1. All four of these `single_key` examples will be exceptions under my new proposal